File: bots/python/example_bot.py
#!/usr/bin/env python3
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import argparse
import json
import threading
import time
import urllib.error
import urllib.request

BOT_NAME = "Python Worker Bot"
BOT_HOST = "127.0.0.1"
BOT_PORT = 8787

session = None
session_lock = threading.Lock()
bot_thread = None
stop_event = threading.Event()


def read_json(handler):
  length = int(handler.headers.get("Content-Length", 0))
  if length == 0:
    return {}
  return json.loads(handler.rfile.read(length).decode("utf-8"))


def post_json(url, payload):
  data = json.dumps(payload).encode("utf-8")
  req = urllib.request.Request(
    url,
    data=data,
    headers={"Content-Type": "application/json"},
    method="POST"
  )
  with urllib.request.urlopen(req, timeout=2) as res:
    return json.loads(res.read().decode("utf-8"))


def get_json(url):
  with urllib.request.urlopen(url, timeout=2) as res:
    return json.loads(res.read().decode("utf-8"))


def action(action_name, args):
  with session_lock:
    current = dict(session) if session else None
  if not current:
    return

  return post_json(f"{current['gameServer']}/api/action", {
    "roomId": current["roomId"],
    "playerId": current["playerId"],
    "action": action_name,
    "args": args
  })


def own_units(state, unit_type=None):
  player_id = state["localPlayerId"]
  units = [
    unit for unit in state.get("units", {}).values()
    if unit["owner"] == player_id and not unit["isMoving"]
  ]
  if unit_type:
    units = [unit for unit in units if unit["type"] == unit_type]
  return units


def neighbors(x, z, grid_size=8, grid_height=None):
  if grid_height is None:
    grid_height = grid_size
  if z % 2 == 0:
    coords = [(x-1, z), (x+1, z), (x-1, z-1), (x, z-1), (x-1, z+1), (x, z+1)]
  else:
    coords = [(x-1, z), (x+1, z), (x, z-1), (x+1, z-1), (x, z+1), (x+1, z+1)]
  return [(nx, nz) for nx, nz in coords if 0 <= nx < grid_size and 0 <= nz < grid_height]


def pick_worker_move(state):
  workers = own_units(state, "worker")
  if not workers:
    return None

  grid = state["grid"]
  for worker in workers:
    for nx, nz in neighbors(worker["x"], worker["z"], state.get("gridWidth", state["gridSize"]), state.get("gridHeight", state["gridSize"])):
      cell = grid[nx][nz]
      if cell["type"] == "resource" and cell["gold"] > 0:
        return worker["id"], nx, nz

  worker = workers[0]
  for nx, nz in neighbors(worker["x"], worker["z"], state.get("gridWidth", state["gridSize"]), state.get("gridHeight", state["gridSize"])):
    if grid[nx][nz]["type"] not in ("base", "obstacle"):
      return worker["id"], nx, nz
  return None


def bot_loop():
  while not stop_event.is_set():
    with session_lock:
      current = dict(session) if session else None

    if not current:
      time.sleep(0.2)
      continue

    try:
      state = get_json(
        f"{current['gameServer']}/api/state?roomId={current['roomId']}&playerId={current['playerId']}&events=0"
      )
      state["localPlayerId"] = current["playerId"]

      player = state["players"][str(current["playerId"])]
      if player["crystals"] >= 50:
        action("build", {"unitType": "worker"})

      move = pick_worker_move(state)
      if move:
        unit_id, x, z = move
        action("move", {"unitIds": [unit_id], "toX": x, "toZ": z})
    except (urllib.error.URLError, TimeoutError, KeyError, json.JSONDecodeError):
      pass

    time.sleep(current.get("pollMs", 200) / 1000)


class BotHandler(BaseHTTPRequestHandler):
  def _send(self, status, payload):
    body = json.dumps(payload).encode("utf-8")
    self.send_response(status)
    self.send_header("Access-Control-Allow-Origin", "*")
    self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
    self.send_header("Access-Control-Allow-Headers", "Content-Type")
    self.send_header("Content-Type", "application/json")
    self.send_header("Content-Length", str(len(body)))
    self.end_headers()
    self.wfile.write(body)

  def do_OPTIONS(self):
    self._send(200, {"ok": True})

  def do_POST(self):
    global session, bot_thread
    payload = read_json(self)

    if payload.get("protocol") != "octagon-rts-bot-v1":
      self._send(400, {"error": "Unsupported bot protocol"})
      return

    if self.path == "/handshake":
      self._send(200, {"ok": True, "name": BOT_NAME})
      return

    if self.path == "/session":
      with session_lock:
        session = {
          "gameServer": payload["gameServer"],
          "roomId": payload["roomId"],
          "playerId": payload["playerId"],
          "pollMs": payload.get("pollMs", 200)
        }
      stop_event.clear()
      if not bot_thread or not bot_thread.is_alive():
        bot_thread = threading.Thread(target=bot_loop, daemon=True)
        bot_thread.start()
      self._send(200, {"ok": True, "name": BOT_NAME})
      return

    if self.path == "/stop":
      with session_lock:
        session = None
      self._send(200, {"ok": True})
      return

    self._send(404, {"error": "Not found"})

  def log_message(self, fmt, *args):
    return


if __name__ == "__main__":
  parser = argparse.ArgumentParser(description="Run the Octagon-RTS example Python bot.")
  parser.add_argument("--host", default=BOT_HOST, help=f"Host/interface to bind. Default: {BOT_HOST}")
  parser.add_argument("--port", type=int, default=BOT_PORT, help=f"Port to bind. Default: {BOT_PORT}")
  parser.add_argument("--name", default=BOT_NAME, help=f"Bot name shown in the lobby. Default: {BOT_NAME}")
  args = parser.parse_args()

  BOT_NAME = args.name
  print(f"{BOT_NAME} listening on http://{args.host}:{args.port}")
  ThreadingHTTPServer((args.host, args.port), BotHandler).serve_forever()

File: bots/javascript/exampleBot.js
#!/usr/bin/env node
const http = require('http');

const BOT_NAME = 'JavaScript Worker Bot';
const BOT_PORT = 8787;

let session = null;
let loopTimer = null;

function send(res, status, payload) {
  const body = JSON.stringify(payload);
  res.writeHead(status, {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(body)
  });
  res.end(body);
}

function readJson(req) {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', chunk => { body += chunk; });
    req.on('end', () => {
      try {
        resolve(body ? JSON.parse(body) : {});
      } catch (err) {
        reject(err);
      }
    });
  });
}

async function getJson(url) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`GET failed: ${res.status}`);
  return res.json();
}

async function action(actionName, args) {
  if (!session) return;
  await fetch(`${session.gameServer}/api/action`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      roomId: session.roomId,
      playerId: session.playerId,
      action: actionName,
      args
    })
  });
}

function neighbors(x, z, gridSize = 8, gridHeight = gridSize) {
  const coords = z % 2 === 0
    ? [[x - 1, z], [x + 1, z], [x - 1, z - 1], [x, z - 1], [x - 1, z + 1], [x, z + 1]]
    : [[x - 1, z], [x + 1, z], [x, z - 1], [x + 1, z - 1], [x, z + 1], [x + 1, z + 1]];
  return coords.filter(([nx, nz]) => nx >= 0 && nx < gridSize && nz >= 0 && nz < gridHeight);
}

function hexDistance(x1, z1, x2, z2) {
  function offsetToCube(col, row) {
    const x = col - Math.floor((row - (row & 1)) / 2);
    const z = row;
    const y = -x - z;
    return [x, y, z];
  }

  const [ax, ay, az] = offsetToCube(x1, z1);
  const [bx, by, bz] = offsetToCube(x2, z2);
  return Math.max(Math.abs(ax - bx), Math.abs(ay - by), Math.abs(az - bz));
}

function isPassable(cell) {
  return cell.type !== 'base' && cell.type !== 'obstacle' && cell.type !== 'resource';
}

function ownUnits(state, unitType) {
  return Object.values(state.units || {}).filter(unit => {
    return unit.owner === session.playerId && !unit.isMoving && (!unitType || unit.type === unitType);
  });
}

function resourceCells(state) {
  const cells = [];
  for (const row of state.grid) {
    for (const cell of row) {
      if (cell.type === 'resource' && cell.gold > 0) {
        cells.push(cell);
      }
    }
  }
  return cells;
}

function adjacentMiningTiles(state, resource) {
  const width = state.gridWidth || state.gridSize;
  const height = state.gridHeight || state.gridSize;
  return neighbors(resource.x, resource.z, width, height).filter(([x, z]) => {
    return isPassable(state.grid[x][z]);
  });
}

function pickWorkerMove(state) {
  const workers = ownUnits(state, 'worker');
  if (workers.length === 0) return null;

  const width = state.gridWidth || state.gridSize;
  const height = state.gridHeight || state.gridSize;
  const resources = resourceCells(state);

  for (const worker of workers) {
    if (resources.some(resource => hexDistance(worker.x, worker.z, resource.x, resource.z) === 1)) {
      continue;
    }

    const targets = resources.flatMap(resource => adjacentMiningTiles(state, resource));
    if (targets.length > 0) {
      const [targetX, targetZ] = targets.reduce((best, tile) => {
        return hexDistance(worker.x, worker.z, tile[0], tile[1]) < hexDistance(worker.x, worker.z, best[0], best[1])
          ? tile
          : best;
      });

      const candidates = neighbors(worker.x, worker.z, width, height).filter(([x, z]) => isPassable(state.grid[x][z]));
      if (candidates.length > 0) {
        const [x, z] = candidates.reduce((best, tile) => {
          return hexDistance(tile[0], tile[1], targetX, targetZ) < hexDistance(best[0], best[1], targetX, targetZ)
            ? tile
            : best;
        });
        return { unitId: worker.id, x, z };
      }
    }
  }

  if (resources.length > 0) return null;

  const worker = workers[0];
  for (const [nx, nz] of neighbors(worker.x, worker.z, width, height)) {
    if (isPassable(state.grid[nx][nz])) {
      return { unitId: worker.id, x: nx, z: nz };
    }
  }
  return null;
}

async function botTick() {
  if (!session) return;

  try {
    const state = await getJson(`${session.gameServer}/api/state?roomId=${session.roomId}&playerId=${session.playerId}&events=0`);
    const player = state.players[String(session.playerId)];

    if (player.crystals >= 50) {
      await action('build', { unitType: 'worker' });
    }

    const move = pickWorkerMove(state);
    if (move) {
      await action('move', { unitIds: [move.unitId], toX: move.x, toZ: move.z });
    }
  } catch (_) {
    // Ignore transient game-server errors while the lobby is starting or stopping.
  }
}

function startLoop() {
  if (loopTimer) clearInterval(loopTimer);
  loopTimer = setInterval(botTick, session.pollMs || 200);
}

const server = http.createServer(async (req, res) => {
  if (req.method === 'OPTIONS') {
    send(res, 200, { ok: true });
    return;
  }

  if (req.method !== 'POST') {
    send(res, 405, { error: 'Method not allowed' });
    return;
  }

  let payload;
  try {
    payload = await readJson(req);
  } catch (_) {
    send(res, 400, { error: 'Invalid JSON' });
    return;
  }

  if (payload.protocol !== 'octagon-rts-bot-v1') {
    send(res, 400, { error: 'Unsupported bot protocol' });
    return;
  }

  if (req.url === '/handshake') {
    send(res, 200, { ok: true, name: BOT_NAME });
    return;
  }

  if (req.url === '/session') {
    session = {
      gameServer: payload.gameServer,
      roomId: payload.roomId,
      playerId: payload.playerId,
      pollMs: payload.pollMs || 200
    };
    startLoop();
    send(res, 200, { ok: true, name: BOT_NAME });
    return;
  }

  if (req.url === '/stop') {
    session = null;
    if (loopTimer) clearInterval(loopTimer);
    loopTimer = null;
    send(res, 200, { ok: true });
    return;
  }

  send(res, 404, { error: 'Not found' });
});

server.listen(BOT_PORT, '127.0.0.1', () => {
  console.log(`${BOT_NAME} listening on http://127.0.0.1:${BOT_PORT}`);
});
